### PR TITLE
Refactor parsing of named expressions

### PR DESCRIPTION
**Pull request by da-woods**
I've tried to rewrite it to largely follow the rules from the most
recent version of the Python LL parser, so avoiding conditional
parameters to the `p_*` functions.

Fixes one part of https://github.com/cython/cython/issues/4595

### CODE REVIEW

File: Cython/Compiler/Parsing.py [hunk 1]
```diff
@@ -131,49 +131,64 @@ def p_lambdef(s, allow_conditional=True):
...
+    #   expected ')', found ':='
+    # message into something a bit more descriptive.
+    # It is close to what the PEG parser does in CPython, where an expression has
+    # a lookahead assertion that it isn't followed by ':='
+    expr = p_test_no_namedexpr_check(s)
+    if s.sy == ':=':
+        s.error("invalid syntax: assignment expression not allowed in this context")
+    return expr
+
+def p_test_no_namedexpr_check(s):
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:150]: Why the word `check` here? Either a named expression is allowed or not, but if it's only being checked for, what does that mean?

File: Cython/Compiler/Parsing.py [hunk 2]
```diff
@@ -131,49 +131,64 @@ def p_lambdef(s, allow_conditional=True):
...
+    #     so look for expression
+    #  2. Otherwise, look for expression
+    lhs = p_test_no_namedexpr_check(s)
     if s.sy == ':=':
         position = s.position()
-        if not allow_assignment_expression:
-            s.error("invalid syntax: assignment expression not allowed in this context")
-        elif not lhs.is_name:
+        if not lhs.is_name:
             s.error("Left-hand side of assignment expression must be an identifier")
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:185]: Would it make sense to turn this into a non-fatal exception? We detected `:=`, so it seems safe to continue parsing from here. It's usually helpful to give more errors at a time to speed up the "fix to compile" cycle.
```suggestion
            s.error("Left-hand side of assignment expression must be an identifier", fatal=False)
```

**Review by scoder — COMMENTED** [review]
Looks like a nice improvement.

File: Cython/Compiler/Parsing.py (hunk 1 above)
Comment (da-woods) [reply Cython/Compiler/Parsing.py:150]: I think I meant "p_test without the `if.sy == ':=': s.error("Invalid syntax...")` check".

I see your point about the name - it isn't the best

File: Cython/Compiler/Parsing.py (hunk 2 above)
Comment (da-woods) [reply Cython/Compiler/Parsing.py]: Probably - it does seem like it should be relatively safe

File: Cython/Compiler/Parsing.py (hunk 1 above)
Comment (da-woods) [reply Cython/Compiler/Parsing.py]: Right. I've renamed to `p_test_allow_walrus_after`. I'm still not convinced that's the best name ever but I think it reads a bit more clearly and I'm genuinely struggling to think of anything better.

**Pull request merged**